### PR TITLE
SVG parser: refactor style parser

### PR DESCRIPTION
**Pull request by rodolforg**

### CODE REVIEW

File: synfig-core/src/modules/mod_svg/svg_parser.cpp [hunk 1]
```diff
@@ -2113,3 +2063,173 @@ get_inkscape_version(const xmlpp::Element* svgNodeElement)
...
+std::string
+Style::get(const std::string &property, std::string default_value) const
+{
+	auto item = data.find(property);
+	if (item == data.end())
+		return default_value;
+	return item->second;
+}
+
+double
```
Comment (ice0) [inline synfig-core/src/modules/mod_svg/svg_parser.cpp:2118]: Could you please add a short description of how this method should work?

File: synfig-core/src/modules/mod_svg/svg_parser.cpp [hunk 2]
```diff
@@ -2113,3 +2063,173 @@ get_inkscape_version(const xmlpp::Element* svgNodeElement)
...
+Style::merge_presentation_attributes(const xmlpp::Element *elem)
+{
+	// presentation attributes: https://www.w3.org/TR/SVG/styling.html#ElementSpecificStyling
+	// this list won't increase in new versions. phew!
+
+	static const std::vector<std::string> circle_ellipse_props {"cx","cy"};
+	// geometry_props is for 'foreignObject', 'image', 'rect', 'svg', 'symbol', and 'use'
+	static const std::vector<std::string> geometry_props {"height", "width", "x", "y"};
+	//static const std::vector<std::string> circle_props {"r"};
+	static const std::vector<std::string> circle_rect_props {"rx", "ry"};
```
Comment (ice0) [inline synfig-core/src/modules/mod_svg/svg_parser.cpp:2152]: > rx, ry | ‘ellipse’ and ‘rect’

Looks like ellipse should be here?

File: synfig-core/src/modules/mod_svg/svg_parser.cpp [hunk 3]
```diff
@@ -2113,3 +2063,173 @@ get_inkscape_version(const xmlpp::Element* svgNodeElement)
...
+		}
+		previous_pos = pos;
+		pos++;
+	}
+}
+
+void
+Style::push(const std::string &property, const std::string &value)
+{
+	if (!value.empty() && value != "inherit" && !property.empty())
```
Comment (ice0) [inline synfig-core/src/modules/mod_svg/svg_parser.cpp:2233]: Am I correct and `inherit` is just ignored now?

**Review by ice0 — CHANGES_REQUESTED** [review]

File: synfig-core/src/modules/mod_svg/svg_parser.cpp (hunk 3 above)
Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:2233]: Yes.
Currently it is too.

In most cases, it is the natural way of CSS style processing. And the current code and my proposal use it by default if not overridden.

My proposal to a Style object have all explicitly defined properties. When you go to a child element, you take the parent style and merge the child properties. 'inherit' would mean 'keep that value from my parent'.

Some properties, however, are non-inherited by default, and `inherit` would force it. https://developer.mozilla.org/en-US/docs/Web/CSS/inheritance#non-inherited_properties I'll inspect later (I hope) if any of SVG needs it.

File: synfig-core/src/modules/mod_svg/svg_parser.cpp (hunk 2 above)
Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:2152]: Ouch.

File: synfig-core/src/modules/mod_svg/svg_parser.cpp [hunk 4]
```diff
@@ -2113,3 +2063,173 @@ get_inkscape_version(const xmlpp::Element* svgNodeElement)
...
+	if (item == data.end())
+		return default_value;
+	return item->second;
+}
+
+double
+Style::compute(const std::string &property, std::string default_value, double reference_value) const
+{
+	std::string value = get(property, default_value);
+	if (value.back() != '%')
```
[the remark this answers is not in the document]
Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:2122]: Theoretically, `default_value` would not be an empty string, but yes, we should guarantee anyway.

File: synfig-core/src/modules/mod_svg/svg_parser.cpp [hunk 5]
```diff
@@ -496,22 +482,23 @@ Svg_parser::parser_layer(const xmlpp::Node* node, xmlpp::Element* root, String p
...
 void
-Svg_parser::parser_rect(const xmlpp::Element* nodeElement,xmlpp::Element* root, const String& fill, const String& fill_opacity, const String& opacity)
+Svg_parser::parser_rect(const xmlpp::Element* nodeElement,xmlpp::Element* root, const Style& style)
 {
 	Glib::ustring rect_id		=nodeElement->get_attribute_value("id");
 	Glib::ustring rect_x		=nodeElement->get_attribute_value("x");
 	Glib::ustring rect_y		=nodeElement->get_attribute_value("y");
 	Glib::ustring rect_width	=nodeElement->get_attribute_value("width");
 	Glib::ustring rect_height	=nodeElement->get_attribute_value("height");
+	Glib::ustring fill       	= style.get("fill", "#000");
```
[the remark this answers is not in the document]
Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:501]: We allow 3 hex digits version too. (See `getRed()`)

File: synfig-core/src/modules/mod_svg/svg_parser.cpp [hunk 6]
```diff
@@ -2113,3 +2063,173 @@ get_inkscape_version(const xmlpp::Element* svgNodeElement)
...
+
+	// merge properties with current style
+	// Normally, we just replace them with new values
+	// However, there are some special cases (SVG 1.1 6.14):
+	//   - relative values will have the same units as the value to which it is relative
+	//      and percentage lengths are relative to current viewport or bounding box dimensions
+	//   - if a 'clip-path' property is specified on an ancestor element, and the current element
+	//       has a 'clip-path' of none, the ancestor's clipping path still applies to current element
+
+	for (auto prop : style.data) {
```
[the remark this answers is not in the document]
Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:2099]: Done

File: synfig-core/src/modules/mod_svg/svg_parser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:2134]: Done.

File: synfig-core/src/modules/mod_svg/svg_parser.cpp (hunk 1 above)
Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:2124]: Done

File: synfig-core/src/modules/mod_svg/svg_parser.cpp (hunk 2 above)
Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp]: Done

**Comment by rodolforg** [comment]
Done your review points and rebased to master :)

File: synfig-core/src/modules/mod_svg/svg_parser.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ice0) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:501]: Ok

File: synfig-core/src/modules/mod_svg/svg_parser.cpp [hunk 7]
```diff
@@ -2113,3 +2063,186 @@ get_inkscape_version(const xmlpp::Element* svgNodeElement)
...
+/// This method retrieve a property value, like get() does, but it returns
+/// its numeric value (and it should be used only for numeric properties.
+///
+/// If property is a numerical number, its value should be parsed and computed.
+/// \param property Property name.
+/// \param default_value The value to be used, if this Style does not have this property explicitly set
+/// \param reference_value If property may be percentual, what is the percentage reference
+/// \return the property value as a real number
+double
+Style::compute(const std::string &property, std::string default_value, double reference_value) const
```
Comment (ice0) [inline synfig-core/src/modules/mod_svg/svg_parser.cpp:2125]: Another question. What value it should return for value = `30%`? 
As it seems to me, there should be 0.3, but as I can see from code there will be 30?

Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp:2125]: Ouch. I didn't create a percentage value test yet here XD

I rewrote this method.

File: synfig-core/src/modules/mod_svg/svg_parser.cpp [hunk 8]
```diff
@@ -2113,3 +2063,185 @@ get_inkscape_version(const xmlpp::Element* svgNodeElement)
...
+
+	try {
+		if (value.back() != '%')
+			return std::stod(value);
+		else
+			return reference_value * std::stod(value) / 100.;
+	} catch(...) {
+	}
+
+	if (value != default_value) {
```
Comment (ice0) [inline synfig-core/src/modules/mod_svg/svg_parser.cpp:2137]: There might still be a problem with the empty default value here. For example, if you push a wrong value into a property and pass an empty default. 

Also if you add a test for empty value here, when code begins to repeat yourself.
Here my proposed change:

https://github.com/ice0/synfig/commit/2e105a38ccd09bcf0014621e12e6bc46543cf545

Comment (rodolforg) [reply synfig-core/src/modules/mod_svg/svg_parser.cpp]: done! Thank you :D

**Comment by ice0** [comment]
Merged. Thank you!
